### ISSUE TITLE
Is there no way not to have to wait ~3 hours for a build?

### ISSUE THREAD

**Issue by kootenpv** [melpa/melpa#4280]
Is there no way for us to trigger a rebuild of a specific package? It would seem a lot more efficient?

**Comment by syohex** [comment]
How about [el-get](https://github.com/dimitri/el-get/) or [Cask](https://github.com/cask/cask/) ? They can handle repository directly.

**Comment by kootenpv** [comment]
I'd prefer a solution for MELPA. Is that not something other people would prefer?

**Comment by vyp** [comment]
I don't think there's any way around this because there's just too many packages. But if you want to test/try things for yourself without having to wait for melpa to catch up, [quelpa](https://github.com/quelpa/quelpa) installs locally just as melpa would. (This was one of the reasons quelpa was created even though el-get already existed.)

**Comment by kootenpv** [comment]
If a person has to trigger a build themselves, I don't think it is going to be too much? Wouldn't it be rather less than having to do a lot of checks?

**Comment by vyp** [comment]
But then I worry that people would abuse it.

And do you mean like some script that polls/checks git repositories for commits and then triggers rebuild of the appropriate package? (It may only work for git packages but that's probably the majority.) I think maybe that could work?

**Comment by kootenpv** [comment]
Not sure, but I know that for example for `pip` it works. What would be the abuse? Perhaps just add some logic that you can only build 10 times a day per github user or whatever.

Another option might be what you suggest. 

Either way, I would like to prevent 3 hour builds; it's not nice for the reputation of emacs/MELPA. It should "go with its time".

**Comment by vyp** [comment]
10/day/user could work well but my reasoning was if it's available for anyone then it would end up being too vulnerable for abuse (like bots or whatever), or otherwise figuring out some permission system might not be worth the hassle. Melpa reviewers already seem to have enough on their hands anyway.

And so personally if I was a melpa server maintainer, I would probably be more comfortable with running some bot/script that I can control to try to semi-automate this instead of allowing some trigger rebuild system. Just my view though, I'm not part of melpa!

And I've found some previous discussion on this topic too btw:

- https://github.com/melpa/melpa/issues/3162
- https://github.com/melpa/melpa/issues/3181 (and I know.. you can see me plugging quelpa here as well! :flushed:)

**Comment by Silex** [comment]
I think the only sane way forward is to change the build system so it changes from something that pulls & builds everything all the time to some daemon that monitors repositories every 15 minutes (or md5sums for mediawiki files) and triggers appropriate rebuilds.

With push notifications that would be easy but you cannot ask for every github developpers to set them up :-) Maybe there's a similar service somewhere that monitors repositories & sends you push notifications.

**Comment by vyp** [comment]
And maybe if not polling all the source repositories simultaneously each time, it could just poll a subset of the repositories each time (so it would crawl through the list).

**Comment by kootenpv** [comment]
Thanks guys, is there perhaps someone from MELPA directly who has some comments as well :)?

**Comment by purcell** [comment]
There are options here, but they all require a decent amount of work. We're polling thousands of repos, and there's no easy solution which will cover all cases. And how much special-casing do we want to do for each of github, gitlab, bitbucket etc?

We could allow people to configure github hooks, for which we would then need to write server-side software to handle the hooks.

We could build in parallel more aggressively, thereby polling and hammering repos more - not great.

The most reasonable option would likely be to less frequently build repos which haven't changed in over a year, but that still leaves a lot to poll, so it won't reduce build time to 5 mins, for example.

**Comment by kootenpv** [comment]
Thanks @purcell. What about having to run a specific command line thing like "melpa register --stable" in a git repo folder? I'm thinking python's `pip` here. I guess they have more resources/servers available, but it should be doable? 

Though I guess it also scores high in the "amount of work" category.

In fact, I can use "register", which takes like 30 seconds, and then I can immediately pull new changes.

**Comment by purcell** [comment]
But then you're authenticated locally, so the server end knows you're the package owner. Our set-up is a bit different: we always build the packages ourselves from your public repo, and often we have no relationship at all with the package authors.

**Comment by alphapapa** [comment]
I think the idea of checking infrequently updated repos less frequently is a good one.  The issue there is that, when one of those repos is finally updated (I came across one earlier today that was just updated for the first time in 3 years), users will have to wait a little while to get the new version.

An idea here is to have a web form on the MELPA site that would allow authors and/or users to requests that packages be added to the next build.  To prevent spam, it could have a form field where the user would paste in the version of the most recent MELPA build, a piece of info that the server should already have locally (not requiring pulling the repo to verify), but that would not be something a spambot would fill in.  In fact, it could be scripted by authors in a git-push hook with curl, making for an informal API.

In fact, if it worked out, maybe the general rule could be changed to build each package once a day, or something like that, and then authors of popular, oft-updated packages could use a git-push hook to post to the form.  Authors like that should be in the loop enough and willing to drop a shell script (which could be standardized) into their `hooks`.

Food for thought.  As Steve said, it requires quite some work, and I can't volunteer to help with that right now, so feel free to ignore my musings.  :)

**Comment by vyp** [comment]
But couldn't a spambot also easily get the most recent melpa build version though?

**Comment by alphapapa** [comment]
Sure.  But why would someone write a bot to do that?  The only thing it would accomplish is to cause package repos to be pulled and checked, and that's what MELPA's already doing now with every package.  Even if someone had a grudge against MELPA, they can't DoS-attack the build system by forcing it to build every package when the build system is already doing that now.  And if someone really wanted to DoS-attack MELPA, they could simply DDoS the web site.  

So the last-version CAPTCHA might not even be necessary.  Even if users and/or authors sometimes used the form to put non-updated packages into the build queue, it would still be building fewer packages every time than it does now.  Seems like there's nothing to lose by setting this up, except the time it would take to create it...  :)

**Comment by vyp** [comment]
Ah I see, so you're saying melpa is only going to build from the build queue only if there's a newer version of the package? That makes sense, although I guess a spambot could just continually keep adding multiple packages to the build queue in an attempt to slow down melpa a little bit (especially if an attacker finds packages with down/nonresponsive upstream urls because that would take a little more time to pull/check).

I would still go for something like periodic polling and asynchronous building because I like how melpa takes care of package building automatically (because this would require a git/vcs hook if you want it built within < 1 day). But to be fair, a git hook or something isn't all that bad either. :)

**Comment by alphapapa** [comment]
> That makes sense, although I guess a spambot could just continually keep adding multiple packages to the build queue in an attempt to slow down melpa a little bit (especially if an attacker finds packages with down/nonresponsive upstream urls because that would take a little more time to pull/check).

Right, but MELPA is already building *every* package *every* time, so if someone did that, it would only be as bad as it is now.  :)

> I would still go for something like periodic polling and asynchronous building because I like how melpa takes care of package building automatically (because this would require a git/vcs hook if you want it built within < 1 day).

Well, some packages are updated very often, like several times a week, and those should obviously be checked very often.  But some haven't been updated for months or years, so I think a 24-hour default delay would be okay there.  :)  And if an author suddenly discovers an important bug that needs fixing, even though it hasn't been updated in a long time, he could always go to the web site or use a git hook to add it to the next build queue.

**Comment by vyp** [comment]
> but MELPA is already building every package every time

Well no, by my understanding isn't it doing it in loops, taking about 3-4 hours each time to build all the packages or so? Which is why the issue was created right. So it could become worse than it is now if someone did that, or I'm misunderstanding?

**Comment by alphapapa** [comment]
I think you are misunderstanding me...

> isn't it doing it in loops, taking about 3-4 hours each time to build all the packages or so?

Right, so if a malicious user added every package to the build queue, it would be the same as it is now.

**Comment by milkypostman** [comment]
We used to have a webserver that had a PHP script to build a package on
demand. Ultimately it was costly (cpu+ maintenance) to run that server and
it wasn't used that much.

Unfortunately I think that most of our time building is spent waiting on
repo pulls and we can't know when to build until we have pulled recent
changes. I'm not sure doing a remote query to see if there is an actual
change would be any faster. By the time you query you might as well
transfer.

**Comment by vyp** [comment]
@alphapapa Ah yes I understand now! Sorry for the confusion. I do think that is a good suggestion! :)

**Comment by Fuco1** [comment]
> With push notifications that would be easy but you cannot ask for every github developpers to set them up :-) Maybe there's a similar service somewhere that monitors repositories & sends you push notifications.

Add OAuth to Melpa then do the setup for the developer.  Look at travis or coveralls, setting up a new project is just clicking one button.  Additionally Melpa would get webhooks calls on each push along with the metadata and can decide to build or not.

Let's be realistic here, 99% of people use GitHub.  Maybe dropping other platforms is a good thing to do for the extra benefits.

**Comment by vyp** [comment]
As much as I use GitHub, I don't think we should limit packages to just be on GitHub.

**Comment by Fuco1** [comment]
@vyp We've discussed it [here](https://github.com/melpa/melpa/issues/4798) and [here](https://github.com/melpa/melpa/pull/4802).  The data is pretty clear about how people (want to) use Melpa.  One thing I value in programming more and more as I "mature" as a programmer is simplicity and pragmatism.  Just as supporting one database engine is enough, just as supporting one vcs/provider used by the 99% is also enough.  We should add more when necessary not when we anticipate it might be useful.

In either case Melpa could still simply poll the sources outside github, which would count about 50 packages, and that can be done every 10 minutes so the build times will be good.

**Comment by milkypostman** [comment]
the best thing for melpa would probably be for us to actually keep track of the repos last state. I would guess that 99% of repos don't actually change but we naively rebuild the package anyways. This is part of the issue with our build loop. Now that we actually have a metadata file, we should just put in the commit ID where we built the package and check that on subsequent builds.

unfortunately, i'm going to be busy with some MELPA servers stuff for a while. When that's done i'll try to carve out time for it.

**Comment by vyp** [comment]
> just supporting one vcs/provider

@Fuco1 so I was mainly concerned over non-GitHub git sources. WDYT for those? I think that being git, you can do pretty much the same stuff with them that you can with GitHub hosted packages, so the overhead for supporting them should be minimal since we're supporting GitHub anyway, and so we should. I agree with your linked issues for the reasons that you said. 👍 I'd probably even reject mercurial packages from now on and ask them to strongly reconsider using git instead. (Now that evil is on GitHub.)

@milkypostman Nice! I think that would be huge for melpa, I've been suggesting that for a while! 👍 👍

**Comment by alphapapa** [comment]
I agree with @Fuco1 that aiming for a simple solution rather than an all-encompassing one is generally good (i.e. the "worse is better" theory).  However, I think we should be careful not to build a system that is inflexible, because nothing lasts forever (except maybe Emacs! haha), and I'm sure that someday we'll want to move beyond GitHub for one reason or another.  Considering everyone's limited time, it would be good if making the necessary changes in such a case didn't require major refactoring.  IOW, we don't need to support everything now, but let's be ready to support other systems in the future.

**Comment by kootenpv** [comment]
@milkypostman @Fuco1 It sounds like that's the solution?

Implementing exponential backoff for the sources on Github, and leave everything else equal... It should be a tremendous speed up :)

**Comment by vermiculus** [comment]
Sorry to bump an old conversation, but I wanted to add a few thoughts:

Taking another direction, are we cloning the entire history? Would it not be sufficient to clone just the latest revision?

https://github.com/melpa/package-build/blob/7733c252691c2c1b8491c375176ca51a6e0dfd33/package-build.el#L250

If we do in the future decide to use 'one true host', then could git mirrors be used?

**Comment by purcell** [comment]
> Taking another direction, are we cloning the entire history? 

Yes.

> Would it not be sufficient to clone just the latest revision?

No, not currently, and I've explored this before: the fundamental issue is that the overall package timestamp is the timestamp of the latest commit that touched one of the included files, and that timestamp can be arbitrarily deep in the repo's history.

**Comment by Silex** [comment]
> No, not currently, and I've explored this before: the fundamental issue is that the overall package timestamp is the timestamp of the latest commit that touched one of the included files, and that timestamp can be arbitrarily deep in the repo's history.

The reason behind this is to avoid uncessary new versions when modifying files that are not part of the distribution?

**Comment by milkypostman** [comment]
Yes

**Comment by Silex** [comment]
How hard would it be to change from a timestamp based system to a `md5sum` based system?

**Comment by milkypostman** [comment]
You need something that is monotonically increasing.

**Comment by vermiculus** [comment]
Could we check the hash of the most recent commit and, if it's changed, clone the entire history again? If one of the hashes changes in the history, then the `master` hash will change as well -- this should allow us to short-circuit the great many packages that don't change from day-to-day.

**Comment by purcell** [comment]
> If one of the hashes changes in the history, then the master hash will change as well -- this should allow us to short-circuit the great many packages that don't change from day-to-day.

I can't imagine us being able to optimise this any better than `git` presumably optimises `git pull`, or am I missing something?

**Comment by Silex** [comment]
> You need something that is monotonically increasing.

I don't see why we need this. If the files changed, rebuild the package, and use the current date as the package version. Currently detecting wether a file is changed is done throught timestamps, but it could be done through hashing (e.g md5sum).

**Comment by vermiculus** [comment]
> That’s a decent idea. But I like that the version is tied to something related to the state of the repo. The problem with using the current date is that if we have to rebuild a package because we deleted the old one. Then the md5 is technically the same but it comes out as a version bump which it’s not. 

Why delete the old one?

**Comment by milkypostman** [comment]
Sometimes we delete a package or move servers or whatever. It happens. It’s just not as robust. It requires us to maintain state which we’re trying to avoid. But this is still a decent idea.

**Comment by purcell** [comment]
I'm a bit wary of builds not being reproducible, tbh. Given a source repo and recipe, I feel like the package produced by `package-build` should be identical, regardless of when or where it is built. I'd worry about downstream tools (quelpa, cask etc) being affected if we dropped this.

**Comment by vermiculus** [comment]
> I can't imagine us being able to optimise this any better than `git` presumably optimises `git pull`, or am I missing something?

Do we in fact `git pull` or do we clone anew?

**Comment by milkypostman** [comment]
We pull.
